### ISSUE TITLE
type checking for dynamically loaded filters, and working with hdf5plugin

### ISSUE THREAD

**Issue by d-sot** [hdmf-dev/hdmf#212]
Loading external filters in hdf5 works in h5py. I tried it with H5DataIO but it gives a type checking error. For lz4 filter example:

```
TypeError: incorrect type for 'compression' (got 'int', expected 'str or bool')
```

The registered filter ID in hdf5 for external filters is an int passed as the compression parameter. Their compression options can be multiple parameters, With h5py it is passed as a tuple.

I came across this looking to use blosc compressors for their performance. They're available on pip recently https://pypi.org/project/hdf5plugin/ . They're also available on HDF5 website as a separate package on their download page.

for:
```
# data is a numpy array
H5DataIO(data, chunks=(4,4,4,4),compression=32001,compression_opts=(0, 0, 0, 0, 1, True, 1) )
```

**Comment by oruebel** [comment]
Adding ``int`` to the allowed types for the compression parameter on H5DataIO here

https://github.com/hdmf-dev/hdmf/blob/8602c509d8a9e7d31b6ce0ef6d003232744a6897/src/hdmf/backends/hdf5/h5_utils.py#L361-L364

should probably fix this issue.

**Comment by d-sot** [comment]
Thanks. Adding just int, it first gave an error:
```
File "hdmf/backends/hdf5/h5_utils.py", line 296, in __init__
    str(self.__iosettings['compression']))
ValueError: 32001 compression not support by this version of h5py.
```
just changing the line 486 https://github.com/hdmf-dev/hdmf/blob/8602c509d8a9e7d31b6ce0ef6d003232744a6897/src/hdmf/backends/hdf5/h5_utils.py#L485-L488 to 
```
    return filter in h5py_filters.encode or filter in hdf5plugin.FILTERS.values()
```
fixed it, and it works now.

**Comment by bendichter** [comment]
Nice work! If we wanted to integrate this we'd have to add `hdf5plugins` as a
requirement.

**Comment by rly** [comment]
It looks like `h5py_filters.encode` does not include dynamically loaded filters. Perhaps we do not need the function `filter_available`, and we can let the h5py library raise an error when trying to create a dataset with an invalid compression value.

**Comment by oruebel** [comment]
> ... we can let the h5py library raise an error when trying to create a dataset with an invalid compression value

The main reason we added ``filter_available`` was to avoid exactly this situation, i.e., to validate parameters before write so that it is easier to identify and fix issues before write, and avoid creation of partial files due to errors on write, because of bad HDF5 I/O settings. While we could remove the checks, I think this is a feature that is useful for most users, whereas custom I/O filters is something that much fewer users will use. 

> If we wanted to integrate this we'd have to add h5plugins as a requirement.

I think it may be simplest to: i) allow int as ``compression`` option ``HDF5IO`` and ii) add a boolean flag to the ``HDF5IO`` constructor (e.g., ``allow_custom_filters``)  to indicate the use of custom filters and bypass the error checking in the constructor.  This fix would have the advantage that we don't need to import ``h5plugins`` and I think this should also allow users to use other custom filters as well. I think these changes are simple-enough that it would be fine to add them. 

More generally (beyond the specific issue raised in the bug report), the use of custom I/O filters is a fairly specialized use-case. As such, I think it is fine for users to extend H5DataIO themselves if they want to use custom I/O filters  that require custom options. One main issue with custom I/O filters is that users without the appropriate custom filters installed will not be able to read the file, which is why we have only been using filters that are installed by default with HDF5 so far.

**Comment by bendichter** [comment]
@oruebel that sounds like a reasonable solution

**Comment by d-sot** [comment]
I implemented a filter availability check using ```h5py.filters``` and without importing ```hdf5plugin``` for review. It can also be easily changed to simply bypass the error checking in the constructor.

**Comment by tbenst** [comment]
Thanks for adding support for BLOSC! `zstd` works for me with `allow_plugin_filters=True` on call to `H5DataIO`. However, when I first tried this without the flag, I was baffled by the inscrutable: `ValueError: 32001 compression not support by this version of h5py.` as I tested h5py and verified it worked. Perhaps best to change the message? https://github.com/hdmf-dev/hdmf/blob/552cf4d086e44ce07cdf181d89b1e76b5856fecf/src/hdmf/backends/hdf5/h5_utils.py#L430

Something like ```%s compression may not be supported by this version of h5py. Use `allow_plugin_filters=True` to bypass this check" % ```
